libdrgn/kdump_reader.c

```diff
@@ -0,0 +1,199 @@
+	        !strcmp(kdump_arch_attr, KDUMP_ARCH_ALPHA) ||
+	        !strcmp(kdump_arch_attr, KDUMP_ARCH_IA64) ||
+	        !strcmp(kdump_arch_attr, KDUMP_ARCH_PPC64) ||
+	        !strcmp(kdump_arch_attr, KDUMP_ARCH_S390X));
+}
+
+struct drgn_error *drgn_kdump_get_arch(kdump_ctx_t *ctx,
+                                       enum drgn_architecture_flags *arch)
+{
+	/*
+	 * We first look in the architecture name and from there we
+	 * we determine  whether the architecture is 32 or 64 bits.
+	 * We could also attempt to guess the endianess that way but
+	 * we once again query the kdump for the rare case of
+	 * bi-endian architectures.
+	 */
+	*arch = 0;
+
+	const char *key = KDUMP_ATTR_ARCH_NAME;
+	kdump_attr_ref_t root;
+	kdump_status ks = kdump_attr_ref(ctx, key, &root);
+	if (ks != KDUMP_OK) {
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "kdump_attr_ref(%s) failed: %s\n",
+					 key, kdump_get_err(ctx));
+	}
+
+	kdump_attr_t attr;
+	ks = kdump_attr_ref_get(ctx, &root, &attr);
+	if (ks != KDUMP_OK) {
+		kdump_attr_unref(ctx, &root);
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "kdump_attr_ref_get(%s) failed: %s\n",
+					 key, kdump_get_err(ctx));
+	}
+
+	switch (attr.type) {
+	case KDUMP_STRING:
+		if (drgn_kdump_is_64bits(attr.val.string))
+			*arch |= DRGN_ARCH_IS_64_BIT;
+		break;
+	default:
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "%s - unexpected attr type: %d\n",
+					 key, attr.type);
+	}
+	kdump_attr_unref(ctx, &root);
+
+	key = KDUMP_ATTR_BYTE_ORDER;
+	ks = kdump_attr_ref(ctx, key, &root);
+	if (ks != KDUMP_OK) {
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "kdump_attr_ref(%s) failed: %s\n",
+					 key, kdump_get_err(ctx));
+	}
+
+	ks = kdump_attr_ref_get(ctx, &root, &attr);
+	if (ks != KDUMP_OK) {
+		kdump_attr_unref(ctx, &root);
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "kdump_attr_ref_get(%s) failed: %s\n",
+					 key, kdump_get_err(ctx));
+	}
+
+	switch (attr.type) {
+	case KDUMP_NUMBER:
+		if (attr.val.number == KDUMP_LITTLE_ENDIAN)
+			*arch |= DRGN_ARCH_IS_LITTLE_ENDIAN;
+		break;
+	default:
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "%s - unexpected attr type: %d\n",
+					 key, attr.type);
+	}
+	kdump_attr_unref(ctx, &root);
+
+	return NULL;
+}
+
+struct drgn_error *drgn_read_kdump(void *buf, uint64_t address, size_t count,
+                                   uint64_t offset, void *arg, bool physical)
+{
+	kdump_ctx_t *ctx = arg;
+	size_t nread = count;
+
+	/* zero-out buffer before doing anything */
+	memset(buf, 0, count);
+
+	kdump_addrspace_t as = (physical) ? KDUMP_KPHYSADDR : KDUMP_KVADDR;
+	kdump_status ks = kdump_read(ctx, as, address, buf, &nread);
+	if (ks != KDUMP_OK) {
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "kdump_read failed: %s",
+					 kdump_get_err(ctx));
+	}
+
+	if (nread != count) {
```

Comment:
Can `kdump_read()` validly return a short read, or would this be a bug in libkdumpfile?

Reply:
You are right, I checked the code and it cannot validly return a short read,
as it has its own loop (snippet below).

from https://github.com/ptesarik/libkdumpfile/blob/3943558af6fca25785faca69078b6e1dee1f4fac/src/kdumpfile/read.c :
```
...
kdump_status
kdump_read(kdump_ctx_t *ctx, kdump_addrspace_t as, kdump_addr_t addr,
            void *buffer, size_t *plength)
{
        kdump_status ret;

        clear_error(ctx);
        rwlock_rdlock(&ctx->shared->lock);
        ret = read_locked(ctx, as, addr, buffer, plength);
        rwlock_unlock(&ctx->shared->lock);
        return ret;
}
...
/**  Internal version of @ref kdump_read
 * @param         ctx      Dump file object.
 * @param[in]     as       Address space of @p addr.
 * @param[in]     addr     Any type of address.
 * @param[out]    buffer   Buffer to receive data.
 * @param[in,out] plength  Length of the buffer.
 * @returns                Error status.
 *
 * Use this function internally if the shared lock is already held
 * (for reading or writing).
                                                                                                                     62,0-1        58%
read_locked(kdump_ctx_t *ctx, kdump_addrspace_t as, kdump_addr_t addr,
            void *buffer, size_t *plength)
{
        struct page_io pio;
        size_t remain;
        kdump_status ret;

        ret = KDUMP_OK;
        remain = *plength;
        while (remain) {
                size_t off, partlen;

                pio.addr.as = as;
                pio.addr.addr = page_align(ctx, addr);
                ret = get_page(ctx, &pio);
                if (ret != KDUMP_OK)
                        break;

                off = addr % get_page_size(ctx);
                partlen = get_page_size(ctx) - off;
                if (partlen > remain)
                        partlen = remain;
                memcpy(buffer, pio.chunk.data + off, partlen);
                put_page(ctx, &pio);
                addr += partlen;
                buffer += partlen;
                remain -= partlen;
        }

        *plength -= remain;
        return ret;
}
...
```